ebi_eva_common_pyutils/spring_properties.py

```diff
@@ -26,90 +28,160 @@ def __init__(self, maven_profile, private_settings_file):
 
-    def _common_properties(self, assembly_accession):
+    @staticmethod
+    def _format(*key_value_maps):
+        all_params = defaultdict(list)
+        for key_value_map in key_value_maps:
+            for key in key_value_map:
+                if key_value_map[key] is not None:
+                    all_params[key.split('.')[0]].append(f'{key}={key_value_map[key]}')
+        lines = []
+        for key_type in all_params:
+            for line in all_params[key_type]:
+                lines.append(line)
+            lines.append('')
+
+        return '\n'.join(lines)
+
+    @staticmethod
+    def _format_str(string, param):
+        if param is None:
+            return None
+        elif not param:
+            return ''
+        else:
+            return string.format(param)
+
+    def _common_properties(self, *, assembly_accession, read_preference, chunk_size=100):
         """Properties common to all Spring pipelines"""
         mongo_host, mongo_user, mongo_pass = get_primary_mongo_creds_for_profile(
             self.maven_profile, self.private_settings_file)
         pg_url, pg_user, pg_pass = get_accession_pg_creds_for_profile(self.maven_profile, self.private_settings_file)
         accession_db = get_properties_from_xml_file(
             self.maven_profile, self.private_settings_file)['eva.accession.mongo.database']
-        return f'''spring.datasource.driver-class-name=org.postgresql.Driver
-spring.datasource.url={pg_url}
-spring.datasource.username={pg_user}
-spring.datasource.password={pg_pass}
-spring.datasource.tomcat.max-active=3
-
-spring.jpa.generate-ddl=true
-
-spring.data.mongodb.host={mongo_host}
-spring.data.mongodb.port=27017
-spring.data.mongodb.database={accession_db}
-spring.data.mongodb.username={mongo_user}
-spring.data.mongodb.password={mongo_pass}
-
-spring.data.mongodb.authentication-database=admin
-mongodb.read-preference=secondaryPreferred
-spring.main.web-application-type=none
-spring.main.allow-bean-definition-overriding=true
-spring.jpa.properties.hibernate.jdbc.lob.non_contextual_creation=true
-
-parameters.chunkSize=1000
-parameters.assemblyAccession={assembly_accession}
-'''
-
-    def _accessioning_properties(self, instance):
+        return {
+            'spring.datasource.driver-class-name': 'org.postgresql.Driver',
+            'spring.datasource.url': pg_url,
+            'spring.datasource.username': pg_user,
+            'spring.datasource.password': pg_pass,
+            'spring.datasource.tomcat.max-active': 3,
+            'spring.jpa.generate-ddl': 'true',
+            'spring.data.mongodb.host': mongo_host,
+            'spring.data.mongodb.port': 27017,
+            'spring.data.mongodb.database': accession_db,
+            'spring.data.mongodb.username': mongo_user,
+            'spring.data.mongodb.password': mongo_pass,
+            'spring.data.mongodb.authentication-database': 'admin',
+            'mongodb.read-preference': read_preference,
```

Comment:
Suggest defaulting to 'primary' (unless we have a reason not to. in which case we can override it in the callsite)...
```suggestion
            self._common_properties(assembly_accession=target_assembly, read_preference='primary',
```

Reply:
Done